File: ConstructTreeFromInorderandPostorder.java
import java.util.HashMap;

/**
 * Definition for a binary tree node.
 */


 /*Approach
 1) we use the same approach which we used from inorder and pre order traversal
 2) In preorder and inorder, we start index from start of preorder array, and in this we start from last element of postorder array
 3) coz we need to traverse left, right and root thus root value will be found as the last element.
 4) we recurse also from right and left to avoid the index from going to -1.
 5) recursion can be used through this way.
 */
 
 
class Solution {
    
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }




    HashMap<Integer,Integer> map;
    int index;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        
        if(inorder.length==0 || inorder==null)
        {
            return null;
        }
        
        map = new HashMap<>();
        index = postorder.length-1;
        for(int i=0;i<inorder.length;i++)
        {
            map.put(inorder[i],i);
        }
        
        return helper(postorder,0,postorder.length-1);
        
        
    }
    
    public TreeNode helper(int [] postorder,int start, int end)
    {
        // base case
        if(start>end)
        {
            return null;
        }
        
        // logic
        
        int rootVal = postorder[index];
        index--;
        TreeNode root = new TreeNode(rootVal);
        int rootIndex = map.get(rootVal);
        
        // recursing through postorder
        root.right = helper(postorder,rootIndex+1,end);
        root.left= helper(postorder,start,rootIndex-1);
        
        return root;
        
        
        
    }
}

File: ConstructTreeFromInorderANdPostorder.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    hashmap={}
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        #optimized solution T.C and S.C is O(N)
        self.idx = len(postorder) - 1
        self.hashmap = {}
        for i in range(0,len(inorder)):
            self.hashmap[inorder[i]]=i
        return self.helper(postorder,inorder,0,len(postorder)-1)
    def helper(self,postorder,inorder,start,end):
        #base
        
        if start>end:
            return None
        rootVal = postorder[self.idx]
        self.idx -=1
        root = TreeNode(rootVal)
        rootIndex = self.hashmap.get(rootVal)
        root.right = self.helper(postorder,inorder,rootIndex+1,end)
        root.left = self.helper(postorder,inorder,start,rootIndex - 1)
        return root 
        
        
        
#         #T.C = O(n^2)
#         #S.C = O(n^2)
#         if len(inorder)==0:
#             return None
#         idx = len(postorder) - 1
#         rootVal = postorder[idx]
#         root = TreeNode(rootVal)
#         index = -1
#         for i in range(0,len(inorder)):
#             if(inorder[i]==rootVal):
#                 index = i
#                 break
#         inorderLeft = inorder[0:index]
#         inorderRight = inorder[index+1:len(inorder)]
#         postorderLeft = postorder[0:index]
#         postorderRight = postorder[index:len(postorder)-1]
#         root.left = self.buildTree(inorderLeft,postorderLeft)
#         root.right = self.buildTree(inorderRight,postorderRight)
#         return root